contracts/Synthswap.sol

```diff
@@ -5,85 +5,295 @@ import "./interfaces/ISynthSwap.sol";
+        address _destSynthAddress,
+        bytes calldata _data
     ) external payable override returns (uint) {
+        (uint amountOut,) = swapOn1inch(_data, false);
 
-        // Make sure to set destReceiver to this contract
-        (bool success, bytes memory returnData) = aggregationRouterV3.delegatecall(_data);
-        require(success, _getRevertMsg(returnData));
-        (uint sUSDAmountOut,) = abi.decode(returnData, (uint, uint));
-        
-        sUSD.approve(address(synthetix), sUSDAmountOut);
-        uint amountReceived = synthetix.exchangeWithTrackingForInitiator(
-            'sUSD', // hardcode source currency key
-            sUSDAmountOut, //source amount
-            destinationSynthCurrencyKey,
-            volumeRewards, 
-            'KWENTA' //tracking code
-        );
-        
-        emit SwapInto(msg.sender, amountReceived);
-        return amountReceived;
+        // if destination synth is NOT sUSD, swap on Synthetix is necessary 
+        if (_destSynthCurrencyKey != sUSD_CURRENCY_KEY) {
+            amountOut = swapOnSynthetix(
+                amountOut, 
+                sUSD_CURRENCY_KEY, 
+                _destSynthCurrencyKey, 
+                address(sUSD)
+            );
+        }
+
+        IERC20(_destSynthAddress).safeTransfer(msg.sender, amountOut);
+  
+        emit SwapInto(msg.sender, amountOut);
+        return amountOut;
     }
 
+    /// @inheritdoc ISynthSwap
     function swapOutOf(
-        bytes32 sourceSynthCurrencyKey,
-        uint sourceAmount, // Make sure synthetix is approved to use this amount
+        bytes32 _sourceSynthCurrencyKey,
+        address _sourceSynthAddress,
+        uint _sourceAmount,
         bytes calldata _data
     ) external override returns (uint) {
+        IERC20(_sourceSynthAddress).safeTransferFrom(msg.sender, address(this), _sourceAmount);
+
+        // if source synth is NOT sUSD, swap on Synthetix is necessary 
+        if (_sourceSynthCurrencyKey != sUSD_CURRENCY_KEY) {
+            swapOnSynthetix(
+                _sourceAmount, 
+                _sourceSynthCurrencyKey, 
+                sUSD_CURRENCY_KEY, 
+                _sourceSynthAddress
+            );
+        }
 
-        IERC20 sourceSynth = IERC20(addressResolver.getSynth(sourceSynthCurrencyKey));
-        sourceSynth.transferFrom(msg.sender, address(this), sourceAmount);
+        (uint amountOut, address dstToken) = swapOn1inch(_data, true);
         
-        sourceSynth.approve(address(synthetix), sourceAmount);
-        // We don't use ForInitiator here because we want the sUSD returned to this contract
-        uint sUSDAmountOut = synthetix.exchangeWithTracking(
-            sourceSynthCurrencyKey, 
-            sourceAmount, 
-            'sUSD',
-            volumeRewards, 
-            'KWENTA' //tracking code
-        );
+        if (dstToken == ETH_ADDRESS) {
+            (bool success, bytes memory result) = msg.sender.call{value: amountOut}("");
```

Comment:
Is there a reason we went with this method of transferring ETH? I thought this one was vulnerable to re-entrancy.

Reply:
I believe this is the preferred way of transferring ETH. At this point, all of the ETH in the contract is being transferred to the msg.sender, so if they attempted a Reentrancy attack, it would not benefit them.